lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -106,1360 +56,800 @@ function Install-Prerequisites {
-            $incompleteContexts.Add($context["relation"])
-            $should_restart = $false
-            continue
-        }
-        foreach ($val in $ctx.Keys) {
-            $mergedContext[$val] = $ctx[$val]
-        }
+    if(!$ctxt['instances_dir']) {
+        $ctxt['instances_dir'] = "$NOVA_DEFAULT_INSTANCES_DIR"
     }
-    Set-IncompleteStatusContext -ContextSet $allContexts -Incomplete $incompleteContexts
-    if(!$mergedContext.Count) {
-        return $false
+
+    foreach($dir in @($ctxt['log_dir'], $ctxt['instances_dir'])) {
+        if (!(Test-Path $dir)) {
+            New-Item -ItemType Directory -Path $dir
+        }
     }
-    $cfg = Invoke-RenderTemplateFromFile -Context $mergedContext -Template $service["template"]
-    Set-Content $service["config"] $cfg
-    return $should_restart
-}
 
-function Get-FallbackNetadapter {
-    $name = Get-MainNetadapter
-    $net = Get-NetAdapter -Name $name
-    return $net
+    return $ctxt
 }
 
-function Get-InterfaceFromConfig {
-    Param(
-        [string]$ConfigOption="data-port",
-        [switch]$MustFindAdapter=$false
-    )
-
-    $nic = $null
-    $DataInterfaceFromConfig = Get-JujuCharmConfig -Scope $ConfigOption
-    Write-JujuInfo "Looking for $DataInterfaceFromConfig"
-    if (!$DataInterfaceFromConfig){
-        if($MustFindAdapter) {
-            Throw "No data-port was specified"
-        }
-        return $null
-    }
-    $byMac = @()
-    $byName = @()
-    $macregex = "^([a-f-A-F0-9]{2}:){5}([a-fA-F0-9]{2})$"
-    foreach ($i in $DataInterfaceFromConfig.Split()){
-        if ($i -match $macregex){
-            $byMac += $i.Replace(":", "-")
-        }else{
-            $byName += $i
-        }
+function Uninstall-Nova {
+    $params = @{
+        'Namespace' = "novahyperv"
+        'Key' = "product_name_installed"
     }
-    if ($byMac.Length){
-        $nicByMac = Get-NetAdapter | Where-Object { $_.MacAddress -in $byMac -and $_.DriverFileName -ne "vmswitch.sys" }
+    $productName = Get-CharmState @params
+    if($productName) {
+        $charmServices = Get-CharmServices
+        $serviceNames = $charmServices.Values | ForEach-Object { $_['service'] }
+        Remove-WindowsServices -Names $serviceNames
+
+        Write-JujuInfo "Uninstalling '$productName'"
+        Uninstall-WindowsProduct -Name $productName
+        Remove-CharmState @params
     }
-    if ($byName.Length){
-        $nicByName = Get-NetAdapter | Where-Object { $_.Name -in $byName }
+
+    if(Test-Path $NOVA_INSTALL_DIR) {
+        Remove-Item -Recurse -Force $NOVA_INSTALL_DIR
     }
-    if ($nicByMac -ne $null) {
-        if ($nicByMac.GetType() -ne [System.Array]) {
-            $nicByMac = @($nicByMac)
-        }
-    } else {
-        $nicByMac = @()
+}
+
+function Start-UpgradeOpenStackVersion {
+    $installedRelease = Get-CharmState -Namespace "novahyperv" -Key "product_release_installed"
+    if(!$installedRelease) {
```

Comment:
If we get here, there is a high chance that we will have something installed. If we do, we should uninstall. If we don't, it may mean that this function failed after uninstalling the previous version, but failed instaling the new, in which case we should retry installing. No need to return early. We can continue, and skip installing if desired version matches current version.